File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/ScienceCalculators.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.*

// Kinematic Calculator (Physics)
data class KinematicResult(val displacement: Double?, val velocity: Double?, val acceleration: Double?, val time: Double?)
class KinematicCalculator {
    // v = u + at
    fun finalVelocity(initial: Double, acceleration: Double, time: Double) = initial + acceleration * time
    // s = ut + 0.5at^2
    fun displacement(initial: Double, acceleration: Double, time: Double) = initial * time + 0.5 * acceleration * time * time
    // v^2 = u^2 + 2as
    fun finalVelocityFromDisplacement(initial: Double, acceleration: Double, displacement: Double) = sqrt(initial * initial + 2 * acceleration * displacement)
}

// Force Calculator (F = ma)
data class ForceResult(val force: Double?, val mass: Double?, val acceleration: Double?)
class ForceCalculator {
    fun forceFromMassAccel(mass: Double, acceleration: Double) = mass * acceleration
    fun massFromForceAccel(force: Double, acceleration: Double) = force / acceleration
    fun accelFromForceMass(force: Double, mass: Double) = force / mass
}

// Momentum Calculator (p = mv)
data class MomentumResult(val momentum: Double, val impulse: Double?)
class MomentumCalculator {
    fun momentum(mass: Double, velocity: Double) = mass * velocity
    fun impulse(force: Double, time: Double) = force * time
    fun finalVelocity(mass: Double, initialV: Double, impulse: Double) = initialV + impulse / mass
}

// Work Energy Calculator
data class WorkEnergyResult(val work: Double?, val kineticEnergy: Double?, val potentialEnergy: Double?)
class WorkEnergyCalculator {
    fun work(force: Double, displacement: Double, angleRadians: Double = 0.0) = force * displacement * cos(angleRadians)
    fun kineticEnergy(mass: Double, velocity: Double) = 0.5 * mass * velocity * velocity
    fun potentialEnergy(mass: Double, height: Double, g: Double = 9.81) = mass * g * height
}

// Molar Mass Calculator (Chemistry)
data class MolarMassResult(val element: String, val atomicMass: Double)
class MolarMassCalculator {
    private val elements = mapOf(
        "H" to 1.008, "He" to 4.003, "Li" to 6.941, "Be" to 9.012, "B" to 10.81, "C" to 12.01,
        "N" to 14.01, "O" to 16.00, "F" to 19.00, "Ne" to 20.18, "Na" to 22.99, "Mg" to 24.31,
        "Al" to 26.98, "Si" to 28.09, "P" to 30.97, "S" to 32.07, "Cl" to 35.45, "Ar" to 39.95,
        "K" to 39.10, "Ca" to 40.08, "Fe" to 55.85, "Cu" to 63.55, "Zn" to 65.38, "Ag" to 107.87,
        "Au" to 196.97, "Pb" to 207.2
    )
    fun getMass(symbol: String): Double? = elements[symbol]
    fun calculateH2O() = 2 * (elements["H"] ?: 0.0) + (elements["O"] ?: 0.0) // 18.016
    fun calculateCO2() = (elements["C"] ?: 0.0) + 2 * (elements["O"] ?: 0.0) // 44.01
    fun calculateNaCl() = (elements["Na"] ?: 0.0) + (elements["Cl"] ?: 0.0) // 58.44
}

// pH Calculator (Chemistry)
data class PHResult(val pH: Double, val pOH: Double, val hConc: Double, val ohConc: Double)
class PHCalculator {
    fun fromHConcentration(hConc: Double): PHResult {
        val pH = -log10(hConc)
        val pOH = 14 - pH
        val ohConc = 10.0.pow(-pOH)
        return PHResult(pH, pOH, hConc, ohConc)
    }
    fun fromPH(pH: Double): PHResult {
        val hConc = 10.0.pow(-pH)
        val pOH = 14 - pH
        val ohConc = 10.0.pow(-pOH)
        return PHResult(pH, pOH, hConc, ohConc)
    }
}

// Standard Deviation Calculator
data class StatsResult(val mean: Double, val variance: Double, val stdDev: Double, val count: Int)
class StandardDeviationCalculator {
    fun calculate(numbers: List<Double>): StatsResult {
        val count = numbers.size
        val mean = numbers.sum() / count
        val variance = numbers.map { (it - mean).pow(2) }.sum() / count
        return StatsResult(mean, variance, sqrt(variance), count)
    }
}

File: Calculator/ios-native/CalculatorHub/Calculators/ScienceCalculators.swift
import Foundation

// Kinematic Calculator
class KinematicCalculator {
    func finalVelocity(initial: Double, acceleration: Double, time: Double) -> Double { initial + acceleration * time }
    func displacement(initial: Double, acceleration: Double, time: Double) -> Double { initial * time + 0.5 * acceleration * time * time }
    func finalVelocityFromDisplacement(initial: Double, acceleration: Double, displacement: Double) -> Double { sqrt(initial * initial + 2 * acceleration * displacement) }
}

// Force Calculator (F = ma)
class ForceCalculator {
    func forceFromMassAccel(mass: Double, acceleration: Double) -> Double { mass * acceleration }
    func massFromForceAccel(force: Double, acceleration: Double) -> Double { force / acceleration }
    func accelFromForceMass(force: Double, mass: Double) -> Double { force / mass }
}

// Momentum Calculator (p = mv)
class MomentumCalculator {
    func momentum(mass: Double, velocity: Double) -> Double { mass * velocity }
    func impulse(force: Double, time: Double) -> Double { force * time }
    func finalVelocity(mass: Double, initialV: Double, impulse: Double) -> Double { initialV + impulse / mass }
}

// Work Energy Calculator
class WorkEnergyCalculator {
    func work(force: Double, displacement: Double, angleRadians: Double = 0) -> Double { force * displacement * cos(angleRadians) }
    func kineticEnergy(mass: Double, velocity: Double) -> Double { 0.5 * mass * velocity * velocity }
    func potentialEnergy(mass: Double, height: Double, g: Double = 9.81) -> Double { mass * g * height }
}

// Molar Mass Calculator
class MolarMassCalculator {
    private let elements: [String: Double] = [
        "H": 1.008, "He": 4.003, "Li": 6.941, "Be": 9.012, "B": 10.81, "C": 12.01,
        "N": 14.01, "O": 16.00, "F": 19.00, "Ne": 20.18, "Na": 22.99, "Mg": 24.31,
        "Al": 26.98, "Si": 28.09, "P": 30.97, "S": 32.07, "Cl": 35.45, "Ar": 39.95,
        "K": 39.10, "Ca": 40.08, "Fe": 55.85, "Cu": 63.55, "Zn": 65.38, "Ag": 107.87,
        "Au": 196.97, "Pb": 207.2
    ]
    func getMass(_ symbol: String) -> Double? { elements[symbol] }
    func calculateH2O() -> Double { 2 * (elements["H"] ?? 0) + (elements["O"] ?? 0) }
    func calculateCO2() -> Double { (elements["C"] ?? 0) + 2 * (elements["O"] ?? 0) }
}

// pH Calculator
struct PHResult { let pH: Double; let pOH: Double; let hConc: Double; let ohConc: Double }
class PHCalculator {
    func fromHConcentration(_ hConc: Double) -> PHResult {
        let pH = -log10(hConc); let pOH = 14 - pH; let ohConc = pow(10, -pOH)
        return PHResult(pH: pH, pOH: pOH, hConc: hConc, ohConc: ohConc)
    }
    func fromPH(_ pH: Double) -> PHResult {
        let hConc = pow(10, -pH); let pOH = 14 - pH; let ohConc = pow(10, -pOH)
        return PHResult(pH: pH, pOH: pOH, hConc: hConc, ohConc: ohConc)
    }
}

// Standard Deviation Calculator
struct StatsResult { let mean: Double; let variance: Double; let stdDev: Double; let count: Int }
class StandardDeviationCalculator {
    func calculate(_ numbers: [Double]) -> StatsResult {
        let count = numbers.count; let mean = numbers.reduce(0, +) / Double(count)
        let variance = numbers.map { pow($0 - mean, 2) }.reduce(0, +) / Double(count)
        return StatsResult(mean: mean, variance: variance, stdDev: sqrt(variance), count: count)
    }
}
